BootloaderCorePkg/Tools/GenExtKeyHashStore.py

```diff
@@ -0,0 +1,91 @@
+# SPDX-License-Identifier: BSD-2-Clause-Patent
+#
+##
+
+import os
+import argparse
+import subprocess
+import sys
+import struct
+import binascii
+import string
+from ctypes import *
+
+from CommonUtility import *
+from BuildUtility import gen_pub_key_hash_store, HASH_USAGE
+
+
+#
+#  HASH USAGE definitions supported
+#
+#  'PUBKEY_CFG_DATA'     : (1<<9)
+#  'PUBKEY_FWU'          : (1<<10)
+#  'PUBKEY_OS'           : (1<<11)
+#  'PUBKEY_CONT_DEF'     : (1<<12)
+#  'PUBKEY_OEM_0'        : (1<<24)
+#  'PUBKEY_OEM_1'        : (1<<25)
+#  'PUBKEY_OEM_2'        : (1<<26)
+#  'PUBKEY_OEM_3'        : (1<<27)
+#  'PUBKEY_OEM_4'        : (1<<28)
+#  'PUBKEY_OEM_5'        : (1<<29)
+#  'PUBKEY_OEM_6'        : (1<<30)
+#  'PUBKEY_OEM_7'        : (1<<31)
+
+#
+#  Sample Key Hash Store Layout file
+#
+#      usage ,                                 KeyFile/Key_ID
+# =========================================================================
+#  (    HASH_USAGE['PUBKEY_CFG_DATA'],     'KEY_ID_CFGDATA_RSA2048' ),
+#  (    HASH_USAGE['PUBKEY_FWU'],          'KEY_ID_FIRMWAREUPDATE_RSA2048'),
+#  (    HASH_USAGE['PUBKEY_CONT_DEF'],     'KEY_ID_CONTAINER_RSA2048'  ),
+#  (    HASH_USAGE['PUBKEY_OS'],           'KEY_ID_OS1_PUBLIC_RSA2048' ),
+#
+#
+
+def main():
+
+    #
+    # Create command line argument parser object
+    #
+
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument('-l',  '--layout', dest='keyh_layout',   type=str, required=True, help='Key hash store layout input file')
+    parser.add_argument('-i',  '--input_dir', dest='input_key_dir',   type=str, default='', help='Input keys directory')
+    parser.add_argument('-k',  '--priv_key', dest='PrivKey', type=str, required=True, help='Key Id or Private RSA 2048/RSA3072 key in PEM format to sign image')
+    parser.add_argument('-a',  '--alg_hash', dest='HashType', type=str, choices=['SHA2_256', 'SHA2_384', 'AUTO'], default='AUTO', help='Hash type for signing. For AUTO hash type will be choosen based on key length')
+    parser.add_argument('-s',  '--sign_scheme', dest='SignScheme', type=str, choices=['RSA_PKCS1', 'RSA_PSS'], default='RSA_PSS', help='Signing Scheme types')
+    parser.add_argument('-svn', dest='svn', type=int, default=0, help='Security version number for key hash store')
+    parser.add_argument('-o',  '--output', dest='out_bin', type=str, required=True, help='Output file for signed image')
+
+    #
+    # Parse command line arguments
+    #
+    args = parser.parse_args()
+
+    layout = get_file_data(args.keyh_layout, 'r')
+    key_list = eval ('[%s]' % layout.replace('\\', '/'))
+
+    if args.HashType == 'AUTO':
+        HashType = adjust_hash_type(args.PrivKey)
+
+    gen_pub_key_hash_store (args.PrivKey, key_list, HashType, args.SignScheme, args.svn, args.input_key_dir, args.out_bin)
+
+    # Firmware update need 4K  align files
+    # Align ouput file for 4K size
+    keyh_file_data = get_file_data(args.out_bin)

```

Comment:
Do you really need to align to 4KB ?  I believe the FWU will still update it correctly as long as the binary is smaller than the space available.

Reply:
Currently we have 4K block aligned check's while getting firmware update info.

@raghavag, 4K size alignment for capsule binary is mandatory?

Reply:
Does not necessarily need to be 4k aligned, please try with the exact size.

Reply:
We have the below check in GetFirmwareUpdateInfo(),

    if (ImageHdr->UpdateImageSize & 0xFFF) {
      DEBUG ((DEBUG_INFO, "capsule payload size is not block aligned!"));
      return EFI_UNSUPPORTED;
    }

Reply:
I just saw the code, if we are updating a component from the flash map, we assume that they are block aligned since all the components are usually 4k aligned. But it is possible to remove this check and update for the size as is.

Reply:
Done. Removed  4K align keyhash binary generation.
Also, removed 4K alignment in Fwupdate with PR #798